### PR TITLE
[TF API] Comparison operator overhaul

### PR DESCRIPTION
**Pull request by rxwei**
## Background

1. Operators like `Tensor.<=(_:_:)` today do not return `Bool`, but return `Tensor<Bool>` instead. Having comparison operators in Swift have element-wise semantics is highly inconsistent with 1) the mental model of comparison operators in Swift, 2) the established convention (see [C++ relational operators](http://www.cplusplus.com/reference/vector/vector/operators/)) and 3) the `Comparable` protocol that comparison operators should be associated with.

2. The existing `VectorNumeric` protocol overlaps with `Numeric` in that they both require arithmetic operators. The only reason `VectorNumeric` does not refine `Numeric` is the lack of `Comparable` conformance.

3. A couple of months ago, we changed `Tensor.==(_:_:)` from returning `Tensor<Bool>` to returning `Bool` and added a `elementsEqual(_:)` for element-wise comparison. `Tensor` was then made `Equatable`.

## Solution

As such, we do a three-step overhaul:
1. Rename existing element-wise comparison operators to well-named functions.
2. Make `Tensor` conform to `Comparable`. Comparison operators are lexicographical.
3. Make `VectorNumeric` refine `Numeric` and remove arithmetic operators with type `(Self, Self) -> Self` from `VectorNumeric`.

## This patch

* 1 and 2 described in "Solution" are included.
* Add API `elementsApproximatelyEqual(_:tolerance:)` for the corresponding TensorFlow op `ApproximateEqual`, since it is nice to have.

### CODE REVIEW

**Comment by rxwei** [comment]
**Note**: I haven't added tests, so please do not trigger CI yet.

**Review by dan-zheng — APPROVED** [review]
LGTM. The design makes sense and is necessary for conforming `Tensor` to `Comparable`.

File: stdlib/public/TensorFlow/Ops.swift [hunk 1]
```diff
@@ -386,49 +388,56 @@ public extension Tensor where Scalar : Equatable {
...
-  /// Performs a logical AND operation element-wise.
-  /// - Note: `&&` supports broadcasting.
+  /// Returns a Boolean value indicating whether the value of the first argument
+  /// is lexicographically approximately equal to that of the second argument.
+  /// Errors no larger than `0.00001` are tolerated.
   @inlinable @inline(__always)
-  static func && (lhs: Tensor, rhs: Tensor) -> Tensor {
-    return Raw.logicalAnd(lhs, rhs)
+  static func ≈ (lhs: Tensor, rhs: Tensor) -> Bool {
+    return lhs.elementsApproximatelyEqual(rhs).all()
```
Comment (lattner) [inline stdlib/public/TensorFlow/Ops.swift]: As you know, I'm not a huge fan of innovating in operators without a carefully thought out and well motivated plan.  I'd strongly prefer that all unicode operators have a directly analogous method so people can choose to use the sugar if they want, but aren't punished too much if they don't.

@stephentyrone will point out that comparing to 0.00001 isn't principled, and can be problematic in practice.  Is this precedented from the TF world?

**Review by lattner — APPROVED** [review]
The patch LGTM, but I'd suggest splitting the unicode operator out to its own separate patch for special consideration.  I don't see a strong motivation to include it right now, is there something I'm missing?

File: stdlib/public/TensorFlow/Ops.swift (hunk 1 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/Ops.swift]: TF `ApproximateEqual` has a default error tolerance which is `0.00001`.

**Comment by rxwei** [comment]
@lattner Sure, I can remove `≈(_:_:)` for now.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
Actually, it's not possible to make `Tensor` conform to `Numeric` today because `Tensor` cannot be `ExpressibleByIntegerLiteral` for complicated reasons.

**Comment by lattner** [comment]
Got it, glad to know it's precedented ... :)

**Comment by stephentyrone** [comment]
@rxwei Can you expand a little bit on the issues with conforming to `ExpressibleByIntegerLiteral` (email would be fine if you don't want to put all the gory details in this PR).

**Comment by rxwei** [comment]
@stephentyrone Yes, the issue is with type checking ambiguity.

`Tensor` has tensor+tensor, scalar+tensor, and tensor+scalar arithmetic operators.

```swift
  static func + (lhs: Tensor, rhs: Scalar) -> Tensor
  static func + (lhs: Tensor, rhs: Tensor) -> Tensor
```

When `Tensor` is `ExpressibleByIntegerLiteral`, then expression `x + 2` cannot be resolved because it can be either tensor+tensor or tensor+scalar.

**Pull request merged**